src/linker/Linker/Driver.cs

```diff
@@ -516,27 +519,86 @@ protected int SetupContext (ILogger customLogger = null)
 
 						continue;
 					case "c":
-						if (!GetStringParam (token, l => context.CoreAction = ParseAssemblyAction (l)))
+						if (!GetStringParam (token, l => action = ParseAssemblyAction (l)))
 							return -1;
 
+						if (action == null)
+							return -1;
+
+						context.CoreAction = action.Value;
 						continue;
 					case "u":
-						if (!GetStringParam (token, l => context.UserAction = ParseAssemblyAction (l)))
+						if (!GetStringParam (token, l => action = ParseAssemblyAction (l)))
 							return -1;
 
-						continue;
-					case "p":
-						if (arguments.Count < 2) {
-							ErrorMissingArgument (token);
+						if (action == null)
 							return -1;
-						}
 
-						AssemblyAction action = ParseAssemblyAction (arguments.Dequeue ());
-						context.Actions[arguments.Dequeue ()] = action;
+						context.UserAction = action.Value;
 						continue;
+					case "p": {
+							if (arguments.Count < 2) {
+								ErrorMissingArgument (token);
+								return -1;
+							}
+
+							action = ParseAssemblyAction (arguments.Dequeue ());
+							if (action == null)
+								return -1;
+
+							string assemblyName = arguments.Dequeue ();
+							if (!IsValidAssemblyName (assemblyName)) {
+								context.LogError ($"Invalid assembly name '{assemblyName}'", 1036);
+								return -1;
+							}
+
+							context.RegisterAssemblyAction (assemblyName, action.Value);
+							continue;
+						}
 					case "t":
 						context.KeepTypeForwarderOnlyAssemblies = true;
 						continue;
+#if FEATURE_ILLINK
+					case "x": {
```

Comment:
Can we reuse the "x" case from monolinker? Could we just remove the call to `GetFiles`, or does monolinker really need the '@descriptor.xml' support?

Reply:
Trying not to touch monolinker code in this PR